src/main/scala/io/iohk/ethereum/vm/PrecompiledContracts.scala

```diff
@@ -93,23 +103,118 @@ object PrecompiledContracts {
     def exec(inputData: ByteString): ByteString =
       sha256(inputData)
 
-    def gas(inputDataSize: UInt256): BigInt =
-      60 + 12 * wordsForBytes(inputDataSize)
+    def gas(inputData: ByteString): BigInt =
+      60 + 12 * wordsForBytes(inputData.size)
   }
 
   object Ripemp160 extends PrecompiledContract {
     def exec(inputData: ByteString): ByteString =
       ByteUtils.padLeft(ripemd160(inputData), 32)
 
-    def gas(inputDataSize: UInt256): BigInt =
-      600 + 120 * wordsForBytes(inputDataSize)
+    def gas(inputData: ByteString): BigInt =
+      600 + 120 * wordsForBytes(inputData.size)
   }
 
   object Identity extends PrecompiledContract {
     def exec(inputData: ByteString): ByteString =
       inputData
 
-    def gas(inputDataSize: UInt256): BigInt =
-      15 + 3 * wordsForBytes(inputDataSize)
+    def gas(inputData: ByteString): BigInt =
+      15 + 3 * wordsForBytes(inputData.size)
+  }
+
+  object ModExp extends PrecompiledContract {
+
+    private val lengthBytes = 32
+    private val totalLengthBytes = 3 * lengthBytes
+    private val GQUADDIVISOR = 20
+
+    def exec(inputData: ByteString): ByteString = {
+      val baseLength = getLength(inputData, 0)
+      val expLength = getLength(inputData, 1)
+      val modLength = getLength(inputData, 2)
+
+      if (baseLength == 0 && modLength == 0)
+        ByteString.empty
+      else {
+        val base = getNumber(inputData, totalLengthBytes, baseLength)
+        val exp = getNumber(inputData, safeAdd(totalLengthBytes, baseLength), expLength)
+        val mod = getNumber(inputData, safeAdd(totalLengthBytes, safeAdd(baseLength, expLength)), modLength)
+
+        if (mod == 0) {
+          ByteString.empty
+        } else {
+          val result = base.modPow(exp, mod)
+          ByteString(ByteUtils.bigIntegerToBytes(result.bigInteger, modLength))
+        }
+      }
+    }
+
+    def gas(inputData: ByteString): BigInt = {
+      val baseLength = getLength(inputData, 0)
+      val expLength = getLength(inputData, 1)
+      val modLength = getLength(inputData, 2)
+
+      val expBytes =
+        inputData.slice(
+          safeAdd(totalLengthBytes, baseLength),
+          safeAdd(safeAdd(totalLengthBytes, baseLength), expLength))
+
+      val multComplexity = getMultComplexity(math.max(baseLength, modLength))
+
+      val adjExpLen = adjExpLength(expBytes, expLength)
+
+      val gasCost = multComplexity * math.max(adjExpLen , 1) / GQUADDIVISOR
+
+      gasCost
```

Comment:
maybe consider removing this assignment and just have last line as `multiComplexity * math.max(adjExpLen , 1) / GQUADDIVISOR` and function renamed ad `gasCost` or something